### PR TITLE
Issue/91/zlibtransport integration

### PR DESCRIPTION
**Pull request by SdtElectronics**
* Add the alternative zlib implementation - muzic
* Add tests for muzic
* Add options to enable muzic
* Add tests for TZlibTransport
* Add options to enable TZlibTransport

### CODE REVIEW

File: lib/muzic/src/zlib_comp.c [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+#include <stdio.h>
+#include <string.h>
+
+#include "uzlib.h"
+#include "zlib.h"
+
+void outbits(struct uzlib_comp* out, unsigned long bits, int nbits);
+
+static void out4bytes(struct uzlib_comp* out,
+                      unsigned char st,
```
Comment (cfriedt) [inline lib/muzic/src/zlib_comp.c:16]: What do these variable names mean? Please use descriptive variable names or document the function.

File: tests/lib/thrift/ThriftTest/prj.conf [hunk 2]
```diff
@@ -12,6 +12,7 @@ CONFIG_NETWORKING=y
 CONFIG_NET_SOCKETS=y
 CONFIG_NET_SOCKETPAIR=y
 CONFIG_HEAP_MEM_POOL_SIZE=16384
+CONFIG_NEWLIB_LIBC_MAX_MAPPED_REGION_SIZE=4194304
```
Comment (cfriedt) [inline tests/lib/thrift/ThriftTest/prj.conf:15]: Since this is a config that is really only used for ZlibTransport, it should be listed in the extra_configs field in test case.yaml

File: tests/lib/muzic/src/integration.c [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+  deflateEnd(deflater);
+  free(deflater);
+  deflater = NULL;
+  inflateEnd(inflater);
+  free(inflater);
+  inflater = NULL;
+}
+
+static void test_deflate_finish() {
+  unsigned char defl_buf[1024];
```
Comment (cfriedt) [inline tests/lib/muzic/src/integration.c:62]: Might not be great to allocate such a large amount on the stack.

Please make static / shared among all test cases.

File: tests/lib/muzic/src/integration.c [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+  zassert_equal(inflater->avail_out, 0, NULL);
+  zassert_equal(inflater->total_out, sizeof(sample), NULL);
+  zassert_equal(inflater->total_in, deflater->total_out, NULL);
+
+  zassert_equal(memcmp(infl_buf, sample, sizeof(sample)), 0, NULL);
+}
+
+static void test_deflate_full_flush() {
+  unsigned char defl_buf[1024];
+  unsigned char infl_buf[sizeof(sample)];
```
Comment (cfriedt) [inline tests/lib/muzic/src/integration.c:96]: Please make static / shared among all test cases.

File: tests/lib/muzic/src/integration.c [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+  deflater->avail_in = sizeof(sample);
+  deflater->next_out = defl_buf;
+  deflater->avail_out = sizeof(defl_buf);
+
+  int res = deflate(deflater, Z_FINISH);
+  zassert_equal(res, Z_BUF_ERROR, NULL);
+}
+
+static void test_deflate_full_flush_frag_out() {
+  unsigned char defl_buf[4096];
```
Comment (cfriedt) [inline tests/lib/muzic/src/integration.c:184]: Way too large for the stack.
Please make static / shared among all test cases.

File: tests/lib/muzic/src/integration.c [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+
+  unsigned int len = 65536;
+  char largeInput[len];
+
+  while (len-- > 0) {
+    largeInput[len] = '1' + (len & 0b111111);
+  }
+
+  unsigned char defl_buf[sizeof(largeInput)];
+  unsigned char infl_buf[sizeof(largeInput)];
```
Comment (cfriedt) [inline tests/lib/muzic/src/integration.c:316]: Please make static / shared among all test cases.

File: tests/lib/muzic/src/integration.c [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+  zassert_equal(inflater->avail_out, 1, NULL);
+  zassert_equal(inflater->total_out, sizeof(sample), NULL);
+  zassert_equal(inflater->total_in, deflater->total_out, NULL);
+
+  zassert_equal(memcmp(infl_buf, sample, sizeof(sample)), 0, NULL);
+}
+
+static void test_inflate_frag_out() {
+  unsigned char defl_buf[1024];
+  unsigned char infl_buf[sizeof(sample) + 1];
```
Comment (cfriedt) [inline tests/lib/muzic/src/integration.c:265]: Please make static / shared among all test cases.

File: tests/lib/muzic/src/integration.c [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+  zassert_equal(inflater->avail_out, 0, NULL);
+  zassert_equal(inflater->total_out, sizeof(sample), NULL);
+  zassert_equal(inflater->total_in, deflater->total_out, NULL);
+
+  zassert_equal(memcmp(infl_buf, sample, sizeof(sample)), 0, NULL);
+}
+
+static void test_inflate_frag_in() {
+  unsigned char defl_buf[1024];
+  unsigned char infl_buf[sizeof(sample) + 1];
```
Comment (cfriedt) [inline tests/lib/muzic/src/integration.c:223]: Please make static / shared among all test cases.

File: tests/lib/muzic/src/integration.c [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+  zassert_equal(inflater->avail_out, 0, NULL);
+  zassert_equal(inflater->total_out, sizeof(sample), NULL);
+  zassert_equal(inflater->total_in, deflater->total_out, NULL);
+
+  zassert_equal(memcmp(infl_buf, sample, sizeof(sample)), 0, NULL);
+}
+
+static void test_deflate_full_flush_frag_in() {
+  unsigned char defl_buf[4096];
+  unsigned char infl_buf[sizeof(sample)];
```
Comment (cfriedt) [inline tests/lib/muzic/src/integration.c:129]: Please make static / shared among all test cases.

File: tests/lib/muzic/src/integration.c [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+
+#include <ztest.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <zlib.h>
+
+const char sample[]
```
Comment (cfriedt) [inline tests/lib/muzic/src/integration.c:15]: Please make this static.

File: tests/lib/muzic/src/integration.c [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+  zassert_equal(inflater->total_in, deflater->total_out, NULL);
+
+  zassert_equal(memcmp(infl_buf, sample, sizeof(sample)), 0, NULL);
+}
+
+/* Output large enough to make the internal buffer of inflate rewind */
+static void test_inflate_buf_rewind() {
+
+  unsigned int len = 65536;
+  char largeInput[len];
```
Comment (cfriedt) [inline tests/lib/muzic/src/integration.c:309]: VLAs are very bad, and you really do not want to allocate this on the stack

File: lib/muzic/src/zlib_comp.c [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+                      unsigned char nd,
+                      unsigned char rd,
+                      unsigned char th) {
+  int newlen = out->outlen + 4;
+  if (newlen > out->outsize) {
```
Comment (cfriedt) [inline lib/muzic/src/zlib_comp.c:21]: This is going to be extremely costly. Is there no way to determine how many bytes might be needed and then make 1 large dynamic allocation?

File: tests/lib/muzic/prj.conf [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+
+CONFIG_MUZIC=y
+
+CONFIG_POSIX_API=y
+CONFIG_HEAP_MEM_POOL_SIZE=65536
+
+# Test dependencies
+CONFIG_ZTEST=y
+
+# default stack sizes are not enough
```
Comment (cfriedt) [inline tests/lib/muzic/prj.conf:13]: The default stack sizes are not enough because they are being smashed in most test functions.

Any time an array is declared at function scope without the `static` keyword, it's using stack memory.

Additionally, using a variable to declare the size of an array inside of a function is called a VLA. They are generally considered very bad.

A better solution would be to make the buffer static at file scope (i.e statically allocated in neither heap or stack memory but in the .data section).

**Review by cfriedt — CHANGES_REQUESTED** [review]

**Comment by cfriedt** [comment]
@SdtElectronics - i see several huge and variable-length arrays declared in function scope in one of the trat files. They should be declared static and at file scope.

There are some compilation failures in CI that need to be addressed as well.

File: lib/muzic/src/zlib_comp.c (hunk 12 above)
Comment (SdtElectronics) [reply lib/muzic/src/zlib_comp.c:21]: This bothered me as well. The underlying deflate implementation comes from uzlib, and it used a dynamically growing buffer for output. The size of this buffer grows linearly instead of exponentially (and what even the worse is it doesn't check if `realloc` succeed or not). I recognized the performance impact but decided to leave it untouched back then, to focus on the compatibility layer with Zlib. It worth some effort to rewrite the output buffer of uzlib's deflate implementation.

Comment (cfriedt) [reply lib/muzic/src/zlib_comp.c:21]: I don't think this is something worth optimizing now either. Functionality first. I would make a note of this in #91 too. Also, please remember to make a note about the 32kB buffer size in #91. Anything that is essentially converted to a deferred problem or "for later optimization" should be explicitly noted (just so it is not forgotten).

Comment (SdtElectronics) [reply lib/muzic/src/zlib_comp.c:21]: Sure. In fact, I have implemented the scalable buffer yesterday, and it doesn't require a 32kB buffer (from the beginning) now. But I think it's better to make it into another PR, as there are possibly some improvements for the current implementation. I'll make a note for this first.

**Comment by cfriedt** [comment]
@SdtElectronics - a few housekeeping changes

1. Please move lib/muzic contents to thrift source dir (the goal was to keep TZlibTransport implementation details private). Conditionally compiled source can be activated with the thrift zlib option, share same CMakeLists.txt, etc. muzic tests can be under lib/thrift/muzic.

2. Please cleanly separate into 2+ commits (1+ for lib and 1+ for tests). It might make sense to keep ThriftTest commit separate from muzic test commit. Please ensure muzic has some basic ztest coverage.

3. Please ensure all tests are Apache licensed.

4. For each 3rd party source file, please add the file in it's original format as a separate commit. In the commit message, please include the URL of the original source and the exact git commit of the file as it was taken. Any changes to the original sources should happen in separate commits.

5. For any 3rd party sources, we need to do a license review. That means we need to verify the file / license in it's original state. If the SPDX identifier was added, it should be done in a separate commit and checked for accuracy.

@stephanosio - can you please help with reviewing 3rd party sources?

**Comment by stephanosio** [comment]
My suggestion on the external code contribution would be to use the format documented by the project:
https://docs.zephyrproject.org/latest/contribute/guidelines.html#identifying-contribution-origin.

See https://github.com/zephyrproject-rtos/zephyr/commit/c96f15ecf30fcf75d1dfd654af49c18f7687a80c for example.

There should be two separate commits for importing these external libraries: the first one "importing" the uzlib as-is, and the second one "importing" the "Inflater" as-is.

Also you might want to specify the MIT and Zlib licences as "additional" licences in the `license_config.yml` so that the scancode action does not report a failure.

Note that both MIT and Zlib are OSI-approved licences and their inclusion into this module should not be a problem for the project.

File: lib/thrift/CMakeLists.txt [hunk 14]
```diff
@@ -11,6 +11,7 @@ target_compile_definitions(app PUBLIC _GLIBCXX_HAS_GTHREADS=1)
 
 zephyr_library()
 zephyr_include_directories(include)
+zephyr_include_directories(muzic/include)
```
Comment (stephanosio) [inline lib/thrift/CMakeLists.txt:14]: Is `muzic` supposed to be available outside this module? (i.e. should you be able to use these headers directly in your app?)

If not, this should use `zephyr_library_include_directories()`.

File: lib/thrift/muzic/include/defl_static.h [hunk 15]
```diff
@@ -30,8 +30,8 @@
...
-void outbits(struct uzlib_comp *ctx, unsigned long bits, int nbits);
-void zlib_start_block(struct uzlib_comp *ctx);
-void zlib_finish_block(struct uzlib_comp *ctx);
-void zlib_literal(struct uzlib_comp *ctx, unsigned char c);
-void zlib_match(struct uzlib_comp *ctx, int distance, int len);
+void outbits(struct uzlib_comp* ctx, unsigned long bits, int nbits);
+void zlib_start_block(struct uzlib_comp* ctx);
+void zlib_finish_block(struct uzlib_comp* ctx);
+void zlib_literal(struct uzlib_comp* ctx, unsigned char c);
+void zlib_match(struct uzlib_comp* ctx, int distance, int len);
```
Comment (stephanosio) [inline lib/thrift/muzic/include/defl_static.h:43]: These style/format changes should be in their own commit; that way, it is easy to identify which "functional" changes are actually required to make this work without all the noises.

File: lib/thrift/muzic/include/defl_static.h [hunk 16]
```diff
@@ -1,3 +1,9 @@
+/*
+ * Copyright 2022 uzlib authors
+ *
+ * SPDX-License-Identifier: Zlib
+ */
```
Comment (stephanosio) [inline lib/thrift/muzic/include/defl_static.h:5]: simply adding `/* SPDX-License-Identifier: Zlib */` should be enough.

File: tests/lib/thrift/muzic/src/integration.c [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+      " quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo conse quat.  "
+      "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat"
+      " nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui    "
+      "officia deserunt mollit anim id est laborum.";
+
+static z_stream* deflater;
+static z_stream* inflater;
+
+static void setup() {
+  deflater = (z_stream*)malloc(sizeof(z_stream));
```
Comment (stephanosio) [inline tests/lib/thrift/muzic/src/integration.c:32]: This is a really strange style choice for C (i.e. not C++) code, but I guess we can tolerate it for this specific module.

File: lib/thrift/muzic/src/defl_static.c [hunk 18]
```diff
@@ -0,0 +1,275 @@
+/*
```
Comment (cfriedt) [inline lib/thrift/muzic/src/defl_static.c:1]: please run `git mv lib/thrift/muzic/src/*.c lib/thrift/src` and `mkdir lib/thrift/include/muzic; git mv lib/thrift/muzic/include/*.h lib/thrift/include/muzic`. 

In relevant source files, you could probably adjust e.g. `#include "uzlib.h"` to `#include <muzic/uzlib.h>`.

**Review by cfriedt — CHANGES_REQUESTED** [review]

File: lib/thrift/CMakeLists.txt [hunk 19]
```diff
@@ -30,3 +31,14 @@ zephyr_library_sources(
   ${THRIFT_UPSTREAM}/lib/cpp/src/thrift/TApplicationException.cpp
   ${THRIFT_UPSTREAM}/lib/cpp/src/thrift/TOutput.cpp
 )
+
+zephyr_library_sources_ifdef(CONFIG_THRIFT_ZLIB
```
Comment (cfriedt) [inline lib/thrift/CMakeLists.txt:35]: Looks like this is missing `_TRANSPORT` at the end

Also, you can likely combine these two `zlib_library_sources_ifdef()` blocks into one.

**Review by cfriedt — APPROVED** [review]
